experimental/split_mlir/src/iree/split_mlir/execution.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import List, Optional
+from .types import OpArguments, Tensor, ExecuteOp, OperationList
+
+
+def collect_arguments(arguments: OpArguments,
+                      results: List[List[Tensor]]) -> List[Tensor]:
+  return [results[arg[0]][arg[1]] for arg in arguments]  # type: ignore
+
+
+def execute_operation_list(
+    input: List[Tensor],
+    operation_list: OperationList,
+    execute_op: ExecuteOp,
+    override_results: Optional[List[List[Tensor]]] = None
+) -> List[List[Tensor]]:
+  """Algorithm to execute a call list.
+
+  Parameters
+  ----------
+  input : Input of the graph.
+  execute_op : Callable that executes an operation from the graph.
+  override_results : When execting operations override arguments with this values,
+    instead of using the computed resuts from previous functions.
+  
+  Returns
+  -------
+  All results from all operations is the graph are in the same order
+  as they appear in `operation_list`. `input` is prepened to the result.
+  """
+  results = [input]
+  for call in operation_list[1:]:
```

Comment:
Assuming I've interpreted the code here properly, is there a reason we're executing each individual op on its own with IREE rather than letting it execute a function like any other model? Trying to understand the motivation for breaking down the IR like this.

Reply:
The use case is to call functions. There are really two things, call a function and return. As you point out this is doing kinda the same thing as an execution of the whole program in IREE. Here the difference is that you get the intermediate results and there is the option to override intermediate results when they are fed to subsequent computations.
I could have achieved this as a MLIR pass to transform the main caller function that calls all the others. There would have been some complications when bisecting inner function in the call graph when hunting for the problematic piece. Another thing is mapping which input comes from which function. There is also the problem of having the functions coexist in one IREE module, which is technically stateful.
Considering that this PR turned out to be 500 lines of code. The other approach may have been better.

Reply:
I may rework it at some point.

Reply:
I forgot to mention the other reason for really splitting them into separate MLIR files is when we do pipeline parallel execution of a sharded graph accross devices. In that scenario there is really no way around it. You will have to run the sub-functions in separate modules.